server/src/main/java/org/elasticsearch/common/settings/LocallyMountedSecrets.java

```diff
@@ -0,0 +1,169 @@
+        "locally_mounted_secrets",
+        a -> new LocalFileSecrets((Map<String, String>) a[0], (ReservedStateVersion) a[1])
+    );
+
+    private final String secretsDir;
+    private final String secretsFile;
+    private final SetOnce<LocalFileSecrets> secrets = new SetOnce<>();
+
+    /**
+     * Direct constructor to be used by the CLI
+     */
+    public LocallyMountedSecrets(Environment environment) {
+        var secretsDirPath = environment.configFile().toAbsolutePath().resolve(SECRETS_DIRECTORY);
+        var secretsFilePath = secretsDirPath.resolve(SECRETS_FILE_NAME);
+        secretsParser.declareObject(ConstructingObjectParser.constructorArg(), (p, c) -> p.map(), SECRETS_FIELD);
+        secretsParser.declareObject(ConstructingObjectParser.constructorArg(), (p, c) -> ReservedStateVersion.parse(p), METADATA_FIELD);
+        if (Files.exists(secretsDirPath) && Files.exists(secretsFilePath)) {
+            try {
+                secrets.set(processSecretsFile(secretsFilePath));
+            } catch (IOException e) {
+                throw new IllegalStateException("Error processing secrets file", e);
+            }
+        } else {
+            secrets.set(new LocalFileSecrets(Map.of(), new ReservedStateVersion(-1L, Version.CURRENT)));
+        }
+        this.secretsDir = secretsDirPath.toString();
+        this.secretsFile = secretsFilePath.toString();
+    }
+
+    /**
+     * Used by {@link org.elasticsearch.bootstrap.ServerArgs} to deserialize the secrets
+     * when they are received by the Elasticsearch process. The ServerCli code serializes
+     * the secrets as part of ServerArgs.
+     */
+    public LocallyMountedSecrets(StreamInput in) throws IOException {
+        this.secretsDir = in.readString();
+        this.secretsFile = in.readString();
+        if (in.readBoolean()) {
+            secrets.set(LocalFileSecrets.readFrom(in));
+        }
+        // TODO: Add support for watching for file changes here.
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeString(secretsDir);
+        out.writeString(secretsFile);
+        if (secrets.get() == null) {
+            out.writeBoolean(false);
+        } else {
+            out.writeBoolean(true);
+            secrets.get().writeTo(out);
+        }
+    }
+
+    @Override
+    public boolean isLoaded() {
+        return secrets.get() != null;
+    }
+
+    @Override
+    public Set<String> getSettingNames() {
+        assert isLoaded();
+        return secrets.get().map().keySet();
+    }
+
+    @Override
+    public SecureString getString(String setting) {
+        assert isLoaded();
+        var value = secrets.get().map().get(setting);
+        if (value == null) {
+            return null;
+        }
+        return new SecureString(value.toCharArray());
+    }
+
+    @Override
+    public InputStream getFile(String setting) throws GeneralSecurityException {
```

Comment:
Would it be worth throwing an exception here? At least we should probably add a comment explaining that this implementation doesn't support files as setting value sources.

Reply:
Good catch, I'll actually add an implementation with ByteArrayInputStream, just like StatelessSecureSettings.